### PR TITLE
OOT framebuffer emulation

### PR DESCRIPTION
**Pull request by DaMarkov**
Changed GLideN64 so that it is compatible with OOT with framebufer emulation being turned on.
Another PR to the OOT project will be necessary to turn framebuffer emulation on.
This PR should be backwards compatible with OOT with framebuffer emulation turned off.

### CODE REVIEW

File: projects/msvc/GLideN64.vcxproj [hunk 1]
```diff
@@ -46,7 +46,7 @@
     <CodeAnalysisRuleSet>AllRules.ruleset</CodeAnalysisRuleSet>
     <CodeAnalysisRules />
     <CodeAnalysisRuleAssemblies />
-    <OutDir>$(SolutionDir)..\..\..\external\$(Platform)\</OutDir>
+    <OutDir>$(MSBuildProjectDirectory)\..\..\..\external\$(Platform)\</OutDir>
     <IntDir>$(SolutionDir)build\$(Platform)\$(Configuration)\$(ProjectName)\</IntDir>
```
Comment (blawar) [inline projects/msvc/GLideN64.vcxproj]: None of the project file changes appear relevant to the PR, and should be removed.

File: src/FrameBuffer.cpp [hunk 2]
```diff
@@ -677,7 +677,8 @@ void FrameBufferList::setBufferChanged(f32 _maxY)
 void FrameBufferList::clearBuffersChanged()
 {
 	gDP.colorImage.changed = FALSE;
-	FrameBuffer * pBuffer = frameBufferList().findBuffer(*REG.VI_ORIGIN & 0xffffff);
+	//FrameBuffer * pBuffer = frameBufferList().findBuffer(*REG.VI_ORIGIN & 0xffffff);
```
Comment (blawar) [inline src/FrameBuffer.cpp]: The original Gliden64 behavior should be supported.  Use the #ifdef NATIVE pre processor flag to signify native or original mode.

File: src/FrameBuffer.h [hunk 3]
```diff
@@ -145,7 +145,7 @@ class FrameBufferList
 
 	static FrameBufferList & get();
 
-private:
+//private:
```
Comment (blawar) [inline src/FrameBuffer.h]: Why was this commented out?

File: src/native/Native.cpp [hunk 4]
```diff
@@ -7,13 +7,18 @@
...
 #include <wchar.h>
 #include "settings.h"
 
-#define START_WIDTH 1280
-#define START_HEIGHT 720
+//#define START_WIDTH 1280
+//#define START_HEIGHT 720
 
-static u64 g_width = START_WIDTH;
+#define START_WIDTH  320
```
Comment (blawar) [inline src/native/Native.cpp]: Why was the start size set to 320 x 240?

File: src/native/Native.cpp [hunk 5]
```diff
@@ -97,7 +102,7 @@ extern "C"
         config.video.windowedWidth = g_width;
         config.video.windowedHeight = g_height;
         dwnd().setWindowSize(g_width, g_height);
-        (*REG.VI_ORIGIN)++;
+        //(*REG.VI_ORIGIN)++;
```
Comment (blawar) [inline src/native/Native.cpp]: Just delete this

File: src/native/Native.cpp [hunk 6]
```diff
@@ -106,6 +111,111 @@ void _CheckInterrupts() {
...
 
+//Copied over from OOT - needs cleanup
+struct Gfx;
+struct OSThread;
+typedef void* OSMesg;
+
+typedef struct OSMesgQueue {
+    /* 0x00 */ OSThread* mtqueue;
+    /* 0x04 */ OSThread* fullqueue;
+    /* 0x08 */ s32 validCount;
```
Comment (blawar) [inline src/native/Native.cpp]: Let’s remove this, and just include the original files directly.  Add the ooot include path to the project.

File: src/native/Native.cpp [hunk 7]
```diff
@@ -133,9 +243,10 @@ extern "C" {
 
         api().RomOpen(romName);
 
-        config.frameBufferEmulation.enable = 0;
-        config.frameBufferEmulation.aspect = Config::aAdjust;
+        config.frameBufferEmulation.enable = 0;//Frame buffer disabled, will be enabled by OOT
+        //config.frameBufferEmulation.aspect = Config::aAdjust;
         
+        config.generalEmulation.hacks |= hack_subscreen | hack_ZeldaMonochrome;
```
Comment (blawar) [inline src/native/Native.cpp]: Did you verify these hacks weren’t already being applied? I thought these hacks were already applied through the rom name logic?

**Review by blawar — CHANGES_REQUESTED** [review]
It looks great so far, thank you!  Only a few small tweaks needed.

**Comment by blawar** [comment]
Also change the default fbe option back to on.  It’s in some config cpp file, where the defaults are set in the constructor.

**Comment by DaMarkov** [comment]
Changed as requested except for including the necessary structs from the OOT project.
When including gfx.h I also get a bunch of macros (like _SHIFTL) from the ultra64 header files which are already defined in GLideN64.
I moved the structs to Native.h.

**Pull request merged**